scripts/metalk8s_cli/mixins/salt.py

```diff
@@ -0,0 +1,119 @@
+import json
+import shlex
+import subprocess
+
+import six
+
+from metalk8s_cli.exceptions import CommandError, CommandInitError
+
+
+class SaltCommandMixin(object):
+    def __init__(self, args):
+        super(SaltCommandMixin, self).__init__(args)
+
+    def run_salt_minion(
+        self, command, local=False, outputter=None, color=True,
+        saltenv=None, pillar=None,
+    ):
+        full_cmd = ['salt-call']
+
+        if local:
+            full_cmd.append('--local')
+        if outputter:
+            full_cmd.extend(['--out', outputter])
+        if color:
+            full_cmd.append('--force-color')
+
+        full_cmd.extend(self._build_args(command))
+
+        if saltenv is not None:
+            full_cmd.append('saltenv={}'.format(saltenv))
+        if pillar is not None:
+            full_cmd.append("pillar='{}'".format(json.dumps(pillar)))
+
+        try:
+            return subprocess.check_output(full_cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as exc:
+            raise CommandError(
```

Comment:
Do we really want to raise here ? For me we may want some time to do something with the output in case of error.

For me we should either:
1- always return a dict with `stdout` `stderr` `retcode`
2- when `retcode != 0` raise something that contain `stderr` `stdout` and `retcode` (I mean a dict containing these informations)

For me 2- look like a good solution most of the time we will just not catch the exception and will get the msg just here but for rare special case we can `try except` and get the stderr stdout, retcode and do what ever we want with this

Reply:
Indeed, I was thinking of implementing the second approach. Using `check_output(cmd, stderr=PIPE)` should do the trick.

Reply:
Now there's a `_run_process` helper, which gives you a `CompletedProcess` namedtuple, with keys `args`, `returncode`, `stdout` and `stderr`. In case `returncode` was not 0, it instead raises a `CommandSubprocessError`, with the same attributes, for callers to catch and log as they want.